src/mscorlib/src/System/Runtime/Loader/AssemblyLoadContext.cs

```diff
@@ -49,38 +65,75 @@ public abstract class AssemblyLoadContext
 #endif // FEATURE_MULTICOREJIT
 
-        protected AssemblyLoadContext()
+        static AssemblyLoadContext()
         {
-            // Initialize the ALC representing non-TPA LoadContext
-            InitializeLoadContext(false);
+            // We register the cleanup of all AssemblyLoadContext that have not been finalized in the AppDomain.ProcessExit
+            AppDomain.CurrentDomain.ProcessExit += OnProcessExit;
         }
 
-        internal AssemblyLoadContext(bool fRepresentsTPALoadContext)
+        protected AssemblyLoadContext() : this(false)
         {
-            // Initialize the ALC representing TPA LoadContext
-            InitializeLoadContext(fRepresentsTPALoadContext);
         }
-        
-        void InitializeLoadContext(bool fRepresentsTPALoadContext)
+
+        internal AssemblyLoadContext(bool fRepresentsTPALoadContext)
         {
-            // Initialize the VM side of AssemblyLoadContext if not already done.
-            GCHandle gchALC = GCHandle.Alloc(this);
-            IntPtr ptrALC = GCHandle.ToIntPtr(gchALC);
-            m_pNativeAssemblyLoadContext = InitializeAssemblyLoadContext(ptrALC, fRepresentsTPALoadContext);
+            // Add this instance to the list of alive ALC
+            lock (assemblyLoadContextAliveList)
+            {
+                if (isProcessExiting)
+                {
+                    // TODO: Add localized version
+                    throw new InvalidOperationException("Cannot create an AssemblyLoadContext when the process is exiting");
+                }
 
-            // Initialize event handlers to be null by default
-            Resolving = null;
-            Unloading = null;
+                // Initialize the VM side of AssemblyLoadContext if not already done.
+                isDefault = fRepresentsTPALoadContext;
+                var thisWeakHandle = GCHandle.Alloc(this, GCHandleType.Weak);
+                var thisWeakHandlePtr = GCHandle.ToIntPtr(thisWeakHandle);
+                m_pNativeAssemblyLoadContext = InitializeAssemblyLoadContext(thisWeakHandlePtr, fRepresentsTPALoadContext);
 
-            // Since unloading an AssemblyLoadContext is not yet implemented, this is a temporary solution to raise the
-            // Unloading event on process exit. Register for the current AppDomain's ProcessExit event, and the handler will in
-            // turn raise the Unloading event.
-            AppDomain.CurrentDomain.ProcessExit += OnProcessExit;
+                // Initialize event handlers to be null by default
+                Resolving = null;
+                Unloading = null;
+
+                assemblyLoadContextAliveList.Add(m_pNativeAssemblyLoadContext, new WeakReference<AssemblyLoadContext>(this));
+            }
+        }
+
+        ~AssemblyLoadContext()
+        {
+            // Remove this instance from the assembly to cleanup on ProcessExit
+            lock (assemblyLoadContextAliveList)
+            {
+                if (isProcessExiting)
```

Comment:
There is no finalization on shutdown in .NET Core - this path should be unreachable.